### PR TITLE
Add new SCRIPT SHOW subcommand to dump script via sha1

### LINKED ISSUE #604: script add dump sub command
ref from: https://github.com/redis/redis/pull/8862

In some scenarios, the business may not be able to find the previously used Lua script and only have a SHA signature. We need a subcommand to dump it for the business. We continue to support this?

@madolson

### PR DESCRIPTION
**Pull request by kukey**
In some scenarios, the business may not be able to find the
previously used Lua script and only have a SHA signature.
Or there are multiple identical evalsha's args in monitor/slowlog,
and admin is not able to distinguish the script body.

Add a new script subcommmand to show the contents of script
given the scripts sha1. Returns a NOSCRIPT error if the script
is not present in the cache.

Usage: `SCRIPT SHOW sha1`
Complexity: `O(1)`

Closes #604.
Doc PR: https://github.com/valkey-io/valkey-doc/pull/143

### CODE REVIEW

File: src/eval.c
Comment (madolson) [inline src/eval.c]: Does this need to be variadic (in that it takes multiple arguments)? You could always just send a multi-exec of individual `SCRIPT DUMP <sha1>` commands. Making a command not-variadic usually makes it easier for users to understand and easier to add more arguments in the future.

File: src/eval.c
Comment (madolson) [inline src/eval.c]: This would ideally be an error instead of just silently continuing, to indicate to the user that they did something wrong. Right now they'll just see the same error as if it wasn't present.

File: src/commands/script-dump.json
Comment (madolson) [inline src/commands/script-dump.json]:
```suggestion
        "since": "8.0.0",
```

File: src/commands/script-dump.json
Comment (madolson) [inline src/commands/script-dump.json]: Probably the wrong command tip. I presume you would want to run this on a single shard, which is the default policy, so this could be omitted.

Comment (kukey) [reply src/commands/script-dump.json]: Yes, the command tip is wrong. Thanks

File: src/commands/script-dump.json
Comment (enjoy-binbin) [inline src/commands/script-dump.json]:
```suggestion
        "arity": 3,
```

File: src/eval.c
Comment (enjoy-binbin) [inline src/eval.c]:
```suggestion
"    Dump a script from the scripts cache.",
```

File: src/eval.c
Comment (enjoy-binbin) [inline src/eval.c]: do we want to return a noscripterr?

File: src/commands/script-dump.json
Comment (enjoy-binbin) [inline src/commands/script-dump.json:16]: maybe we also need to add a command_tips?

**Review by enjoy-binbin — COMMENTED** [review]
please also update the top comment instead of directly linking to an issue

File: src/eval.c
Comment (kukey) [reply src/eval.c]: I think it would be better to return an error, as users can intuitively know what the error is

**Comment by madolson** [comment]
@valkey-io/core-team Vote on adding this new command which allows dumping a specific command. There is still some minor open comment about error wording, but other than that it is broadly updated.

**Comment by madolson** [comment]
@kukey Btw, you need to regenerate the commands.def and check it in.

File: src/eval.c
Comment (zuiderkwast) [reply src/eval.c]: The -ERR vs null reply is a little inconsistent here.

We validate the length here, but we still accept invalid SHA1 strings. For strings with 40 characters other than [0-9a-f], we just return null.

Either we catch all invalid SHA1 values or we shouldn't try to catch them at all.

An alternative is that we do like EVALSHA, we return a -NOSCRIPT error for non-existing scripts and also for invalide SHA1 values. It's kind of the same error.

Comment (enjoy-binbin) [reply src/eval.c]: > An alternative is that we do like EVALSHA, we return a -NOSCRIPT error for non-existing scripts and also for invalide SHA1 values. It's kind of the same error.

yean, i prefer to return noscript

Comment (kukey) [reply src/eval.c]: Got it

**Review by madolson — APPROVED** [review]
Looks OK to me. Waiting on other for core team approval before merging.

**Comment by madolson** [comment]
Once we have directional approval for the merge, please create a PR to the doc repo similar to https://github.com/valkey-io/valkey-doc/blob/main/commands/script-exists.md.

File: src/commands/script-dump.json
Comment (enjoy-binbin) [inline src/commands/script-dump.json:29]: we need to update this?

Comment (enjoy-binbin) [inline src/commands/script-dump.json]: we need to update this

File: src/eval.c
Comment (enjoy-binbin) [inline src/eval.c:741]: i guess we can also keep the old 40 fast error path

**Review by enjoy-binbin — APPROVED** [review]

File: src/eval.c
Comment (kukey) [reply src/eval.c:741]: ok

**Comment by enjoy-binbin** [comment]
@valkey-io/core-team I think this one is ready to merge, i updated the top comment, please take a look.

File: src/commands/script-dump.json
Comment (zuiderkwast) [inline src/commands/script-dump.json]: This is not how we normally describe an error reply.

I think we just don't describe error replies in the reply schema. They're just mentioned in the docs if I remember correctly.

Comment (kukey) [reply src/commands/script-dump.json]: Got it

**Comment by soloestoy** [comment]
I like this idea, this can help users to debug. But the subcommand's name `SCRIPT DUMP` is a bit wired to me, since other commands like `DUMP` and `FUNCTION DUMP`'s reply is rdb format, but this subcommand's reply is just raw. Maybe `SCRIPT GET` or `SCRIPT FETCH` is better.

**Comment by madolson** [comment]
> I like this idea, this can help users to debug. But the subcommand's name SCRIPT DUMP is a bit wired to me, since other commands like DUMP and FUNCTION DUMP's reply is rdb format, but this subcommand's reply is just raw. Maybe SCRIPT GET or SCRIPT FETCH is better.

I don't feel strongly. `SCRIPT GET` would be my preferred option of the two. I guess I would prefer something like `SCRIPT DISPLAY`.

**Comment by zuiderkwast** [comment]
Regarding subcommand name, how about `SCRIPT SHOW`?

Compare: In MySQL it's possible to get the CREATE TABLE command for a table using `SHOW CREATE TABLE mytable`.

**Comment by madolson** [comment]
> Regarding subcommand name, how about SCRIPT SHOW?

I considered that but it just sort of sounds weird to me, but I would prefer that over `SCRIPT GET` and `SCRIPT FETCH`.

File: src/eval.c
Comment (hwware) [inline src/eval.c:737]: Here if you reference script exists command how to check the input parameter, you will see the c->argv[2]->ptr length does not been checked. So my opinion is that if you want to check the length of argv[2]->ptr, you had better prompt a much clearer information to the client, such as "sha1 length is not correct"

**Comment by hwware** [comment]
> > Regarding subcommand name, how about SCRIPT SHOW?
> 
> I considered that but it just sort of sounds weird to me, but I would prefer that over `SCRIPT GET` and `SCRIPT FETCH`.

script fetch is much more officially.

**Comment by zuiderkwast** [comment]
Fetch sounds like "go and get it from somewhere", like from another node or file or something. Another idea: `SCRIPT SOURCE`.

File: src/eval.c
Comment (madolson) [reply src/eval.c:737]: This was discussed in an earlier revision, and I believe we ended on this. Unless you feel strongly I think we should leave it as is.

Comment (zuiderkwast) [reply src/eval.c:737]: I vote to leave it as is.

This is just a fast path to return error. Not a real validation. If we want a real validation, we should also check for invalid characters in the SHA1. Let's not do that.

**Comment by kukey** [comment]
> Once we have directional approval for the merge, please create a PR to the doc repo similar to https://github.com/valkey-io/valkey-doc/blob/main/commands/script-exists.md.

@madolson  https://github.com/valkey-io/valkey-doc/pull/143

**Comment by kukey** [comment]
I prefer `SCRIPT GET`, it's clear and simple.  @valkey-io/core-team which subcommand we select? Or also continue use `SCRIPT DUMP`？

**Comment by PingXie** [comment]
> Regarding subcommand name, how about `SCRIPT SHOW`?

my vote ^

**Comment by madolson** [comment]
@kukey We have an online syncup tomorrow, I think we can quickly get consensus since people's opinions seem all over the place online.

**Comment by madolson** [comment]
@kukey `SCRIPT SHOW` Seems to be the winner from our online discussion.

**Comment by kukey** [comment]
> @kukey `SCRIPT SHOW` Seems to be the winner from our online discussion.

copy

File: src/commands/script-show.json
Comment (zuiderkwast) [inline src/commands/script-show.json]:
```suggestion
        "summary": "Show server-side Lua script in the script cache.",
```

File: src/commands/script-show.json
Comment (zuiderkwast) [inline src/commands/script-show.json]:
```suggestion
            "description": "Lua script if sha1 hash exists in script cache.",
```

**Review by zuiderkwast — APPROVED** [review]

**Comment by madolson** [comment]
@zuiderkwast I've been using the `needs-doc-pr` to just track if it's been opened. The doc PR already exists here: https://github.com/valkey-io/valkey-doc/pull/143. Is this inline with what you were thinking?

**Comment by zuiderkwast** [comment]
OK, sounds good. I didn't notice we remove it before. In redis it was never removed I think.

I missed the link to the doc PR. Can we add the link in the PR description to make it more visible?

**Review by hwware — APPROVED** [review]

File: src/commands.def
Comment (madolson) [inline src/commands.def]:
```suggestion
{MAKE_CMD("show","Show server-side Lua script in the script cache.","O(1).","8.0.0",CMD_DOC_NONE,NULL,NULL,"scripting",COMMAND_GROUP_SCRIPTING,SCRIPT_SHOW_History,0,SCRIPT_SHOW_Tips,0,scriptCommand,3,CMD_NOSCRIPT,ACL_CATEGORY_SCRIPTING,SCRIPT_SHOW_Keyspecs,0,NULL,1),.args=SCRIPT_SHOW_Args},
```

**Pull request merged**
